File: Appendix_A/IoTBackgroundAppVB/StartupTask.vb
Imports Windows.ApplicationModel.Background
Imports Windows.Devices.Gpio

Public NotInheritable Class StartupTask
    Implements IBackgroundTask

    Private Const gpioPinNumber = 5
    Private Const msShineDuration = 4000

    Public Sub Run(taskInstance As IBackgroundTaskInstance) Implements IBackgroundTask.Run
        BlinkLed(gpioPinNumber, msShineDuration)
    End Sub

    Private Function ConfigureGpioPin(pinNumber As Integer) As GpioPin
        Dim gpioControl = GpioController.GetDefault()
        Dim pin As GpioPin = Nothing

        If gpioControl IsNot Nothing Then
            pin = gpioControl.OpenPin(pinNumber)

            If pin IsNot Nothing Then
                pin.SetDriveMode(GpioPinDriveMode.Output)
            End If
        End If

        Return pin
    End Function

    Private Sub BlinkLed(gpioPinNumber As Integer, msShineDuration As Integer)
        Dim ledGpioPin = ConfigureGpioPin(gpioPinNumber)

        If ledGpioPin IsNot Nothing Then
            While True
                SwitchGpioPin(ledGpioPin)
                Task.Delay(msShineDuration).Wait()
            End While
        End If
    End Sub

    Private Sub SwitchGpioPin(gpioPin As GpioPin)
        Dim currentPinValue = gpioPin.Read()
        Dim newPinValue = InvertGpioPinValue(currentPinValue)

        gpioPin.Write(newPinValue)
    End Sub

    Private Function InvertGpioPinValue(currentPinValue As GpioPinValue) As GpioPinValue
        Dim invertedGpioPinValue As GpioPinValue

        If currentPinValue = GpioPinValue.High Then
            invertedGpioPinValue = GpioPinValue.Low
        Else
            invertedGpioPinValue = GpioPinValue.High
        End If

        Return invertedGpioPinValue
    End Function
End Class

File: Chapter03/IoTBackgroundAppCpp/StartupTask.cpp
#include "pch.h"
#include "StartupTask.h"

using namespace IoTBackgroundAppCpp;
using namespace Platform;
using namespace Windows::ApplicationModel::Background;

// The Background Application template is documented at http://go.microsoft.com/fwlink/?LinkID=533884&clcid=0x409

void StartupTask::Run(IBackgroundTaskInstance^ taskInstance)
{
    // 
    // TODO: Insert code to perform background work
    //
    // If you start any asynchronous methods here, prevent the task
    // from closing prematurely by using BackgroundTaskDeferral as
    // described in http://aka.ms/backgroundtaskdeferral
    //

	BlinkLed(pinNumber, msShineDuration);
}

GpioPin^ StartupTask::ConfigureGpioPin(int pinNumber)
{
	auto gpioController = GpioController::GetDefault();

	GpioPin^ pin = nullptr;

	if (gpioController != nullptr)
	{
		pin = gpioController->OpenPin(pinNumber);

		if (pin != nullptr)
		{
			pin->SetDriveMode(GpioPinDriveMode::Output);
		}
	}

	return pin;
}

void StartupTask::BlinkLed(int ledPinNumber, int msShineDuration)
{
	GpioPin^ ledGpioPin = ConfigureGpioPin(ledPinNumber);

	if (ledGpioPin != nullptr)
	{
		while (true)
		{
			SwitchGpioPin(ledGpioPin);

			Sleep(msShineDuration);
		}
	}
}

void StartupTask::SwitchGpioPin(GpioPin^ gpioPin)
{
	auto currentPinValue = gpioPin->Read();

	GpioPinValue newPinValue = InvertGpioPinValue(currentPinValue);

	gpioPin->Write(newPinValue);
}

GpioPinValue StartupTask::InvertGpioPinValue(GpioPinValue currentPinValue)
{
	GpioPinValue invertedGpioPinValue;

	if (currentPinValue == GpioPinValue::High)
	{
		invertedGpioPinValue = GpioPinValue::Low;
	}
	else
	{
		invertedGpioPinValue = GpioPinValue::High;
	}

	return invertedGpioPinValue;
}
